File: DSA10003.cpp
#include <iostream>
#include <vector>
using namespace std;

class DisjointSet {
private:
    vector<int> parent, rank;
public:
    DisjointSet(int n) : parent(n + 1), rank(n + 1, 0) {
        for (int i = 0; i <= n; ++i)
            parent[i] = i;
    }

    int find(int u) {
        if (u != parent[u])
            parent[u] = find(parent[u]);
        return parent[u];
    }

    void unionSet(int u, int v) {
        int x = find(u);
        int y = find(v);

        if (x == y) return;

        if (rank[x] < rank[y]) {
            parent[x] = y;
        } else if (rank[x] > rank[y]) {
            parent[y] = x;
        } else {
            parent[x] = y;
            rank[y]++;
        }
    }
};

bool hasCycle(int v, vector<pair<int, int>>& edges) {
    DisjointSet djs(v);
    for (auto& edge : edges) {
        if (djs.find(edge.first) != djs.find(edge.second)) {
            djs.unionSet(edge.first, edge.second);
        } else {
            return true;
        }
    }
    return false;
}

int main() {
    int T, v, e, a, b;
    cin >> T;
    while (T--) {
        cin >> v >> e;
        vector<pair<int, int>> edges;
        for (int i = 0; i < e; ++i) {
            cin >> a >> b;
            edges.push_back({a, b});
        }
        cout << (hasCycle(v, edges) ? "YES" : "NO") << endl;
    }
    return 0;
}

File: DSA10003.py
class DisjointSet:
    def __init__(self, n=0):
        self.n = n
        self.parent = [i for i in range(n + 1)]
        self.rank = [0] * (n + 1)

    def find(self, u):
        if u != self.parent[u]:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        x = self.find(u)
        y = self.find(v)

        if x == y:
            return

        if self.rank[x] < self.rank[y]:
            self.parent[x] = y
        elif self.rank[x] > self.rank[y]:
            self.parent[y] = x
        else:
            self.parent[x] = y
            self.rank[y] += 1

def function():
    v, e = map(int, input().split())

    ls = list(map(int, input().split()))
    djs = DisjointSet(v)
    cycle = False
    for i in range(0, 2 * v, 2):
        if djs.find(ls[i]) != djs.find(ls[i + 1]):
            djs.union(ls[i], ls[i + 1])
        else:
            cycle = True

    print("YES" if cycle else "NO")
if __name__ == "__main__":
    for i in range(int(input())):
        function()